src/java/org/apache/cassandra/transport/CBUtil.java

```diff
@@ -478,7 +487,32 @@ public static void writeValue(ByteBuffer bytes, ByteBuf cb)
         cb.writeInt(remaining);
 
         if (remaining > 0)
-            cb.writeBytes(bytes.duplicate());
+            addBytes(bytes, cb);
+    }
+
+    public static void addBytes(ByteBuffer src, ByteBuf dest)
+    {
+        if (src.remaining() == 0)
+            return;
+
+        int length = src.remaining();
+
+        //heap buffers are copied this way in order to avoid JVM crashing
```

Comment:
please give more detail, this isn't enough for anyone to understand why this code is this way.  You should also move it within the `src.hasArray` block as that comment only makes sense there.